### ISSUE TITLE
Proposal: don't allow appending enforceBranches on repo level if disableRepoOverride is true

### ISSUE THREAD

**Issue by yorinasub17** [ossf/allstar#211]
Right now, `allstar` always appends `enforceBranches` from the repo level config regardless of the `disableRepoOverride` setting. This is different from all the other policies, where `disableRepoOverride` means that nothing from the repo level is merged in.

I suspect the use case for this is to support different repo branch configurations, where some repos may have additional branches that need to be protected (e.g., if they are using release branches).

However, with the introduction of repo level overrides in the org repo, I think a better posture is to configure the individual repo `enforceBranches` in the `repo/branch_protection.yaml` config in the org repo if an org wants to set `disableRepoOverride = true`.

What is the community's thoughts on removing this feature such that `disableRepoOverride` means the config in each individual repo's `.allstar` folder is always ignored?

## Motivation

My primary motivation for proposing this is to further reduce API calls. This feature is the one blocker in proposing an improvement to Allstar to avoid fetching the repo level config when `disableRepoOverride = true`. If we remove this feature, then we know the repo level config is never merged in, and thus Allstar can avoid looking for the `.allstar` folder in each individual repo when `disableRepoOverride = true` in the org config.

This is less of a problem now that https://github.com/ossf/allstar/pull/197 is implemented, but I am nervous that individual repo maintainers can still potentially DoS Allstar by adding a `.allstar` folder in each repo. This can 5x the number of API calls by adding each policy config file, which gets us closer to reaching API rate limits (thus DoS). Note the cache doesn't help much here since a get response on an individual file is only cached for 60 seconds, and with lots of repos, each repeated call for a repo is more than that (for our org, it is roughly ~10 minutes).

Hence, would love it if we can avoid the repo level config fetching altogether if it is not necessary.

**Comment by yorinasub17** [comment]
For reference, the proposal I am making is to move the merge operation to within the if block here:

https://github.com/ossf/allstar/blob/8cc89a1/pkg/policies/branch/branch.go#L644-L647

**Comment by jeffmendoza** [comment]
> I suspect the use case for this is to support different repo branch configurations, where some repos may have additional branches that need to be protected (e.g., if they are using release branches).

Yes, This was the original intention. Agree that repo level overrides in the org repo can make this obsolete.

> What is the community's thoughts...

Yes, let's see if anyone has a concern.

About the cache. Even if the local cache expires and we make a call out to github, doesn't the "conditional request" feature cause quota to be preserved? Are you looking to avoid calls in general, and not just quota?

**Comment by yorinasub17** [comment]
> Even if the local cache expires and we make a call out to github, doesn't the "conditional request" feature cause quota to be preserved?

So I think you are right about this, but in my observations with request logging (checking quota usage returned in the headers), this does not seem to be the case and the quota was being eaten up pretty rapidly. Let me investigate this further to see if I can trace down why this is occurring and will report back.

**Comment by yorinasub17** [comment]
> Are you looking to avoid calls in general, and not just quota?

FWIW, I am generally a fan of avoiding unnecessary network calls regardless of quota so that we are good netizens in the community 😄
